### PR TITLE
BERT encoder module

### PR DESCRIPTION
**Pull request by iftenney**
BERT support (#441). Implementation modeled after that of the OpenAI transformer model, albeit somewhat cleaner using the excellent `pytorch_pretrained_bert` package (https://github.com/huggingface/pytorch-pretrained-BERT).

Currently works for edge probing experiments; see #460. For other tasks, you'll need to apply the BERT WPM tokenization either offline or at data loading time.

### CODE REVIEW

**Comment by iftenney** [comment]
Note: this also depends on #451

File: config/defaults.conf [hunk 1]
```diff
@@ -198,6 +198,26 @@ openai_transformer_fine_tune = 0  // If false, OpenAI Transformer weights will
                                   // Note that this might require a
                                   // high amount of GPU memory, so consider
                                   // lowering the batch size if set to true.
+bert_model_name = ""  // If nonempty, use this BERT model for representations.
+                      // Available values are:
+                      // bert-base-uncased, bert-large-cased, ...
+                      // For more, see:
+                      // https://github.com/huggingface/pytorch-pretrained-BERT#loading-google-ais-pre-trained-weigths-and-pytorch-dump
+                      // Not compatible with other embeddings due to different
```
[the remark this answers is not in the document]
Comment (iftenney) [reply config/defaults.conf]: Yup.

**Comment by pitrack** [comment]
I'm confused by tokenization. Earlier, each model corresponded to a tokenization scheme (e.g. openai-> openai-bpe). However, now it's broken down by task because bert cases for ner.

Then the order of priority is something like: default is MosesTokenizer(), then use the one specified by the model, then by the task? Eventually, we would want to update the readme or add comments somewhere which suggests this. It also feels a bit redundant (having both tasks have a class variable and another one in args).

File: src/bert/utils.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import os
+import random
+import copy
+import logging as log
+
+from typing import List, Dict
+
+import numpy as np
```
Comment (pitrack) [inline src/bert/utils.py]: [nit] some of these imports aren't used -- e.g. `os`, `copy`, `List`, `numpy`, or `random`. We probably have many more cases of this in the codebase. Perhaps the one to be most wary of is where we call/import `random`

Comment (iftenney) [reply src/bert/utils.py]: Yeah, some of this was copypasta from the openai model wrapper, which requires a lot more preprocessing. Fixed.

File: src/bert/utils.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        Returns:
+            h: [batch_size, seq_len, d_emb]
+        """
+        assert "bert_wpm_pretokenized" in sent
+        # <int32> [batch_size, var_seq_len]
+        var_ids = sent["bert_wpm_pretokenized"]
+
+        # Model has fixed, learned positional component, so we must pass a
+        # block of exactly seq_len length.
```
Comment (W4ngatang) [inline src/bert/utils.py]: Since (I think) the model will always load positional embeddings for 512(?) positions, does `seq_len` need to be set to 512 when using a BERT model?

File: src/bert/utils.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        mask = (ids == 0)
+        # "Correct" ids to account for different indexing between BERT and
+        # AllenNLP.
+        # The AllenNLP indexer adds a '@@UNKNOWN@@' token to the
+        # beginning of the vocabulary, *and* treats that as index 1 (index 0 is
+        # reserved for padding).
+        FILL_ID = 0  # [PAD] for BERT models.
+        ids[ids == 0] = FILL_ID + 2
+        ids -= 2
```
Comment (W4ngatang) [inline src/bert/utils.py]: this will map elements of `ids` that are 1 to -1, potentially causing indexing errors; is there a guard against this or a reason it won't happen?

File: src/bert/utils.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        if self.embeddings_mode != "none":
+            # Extract lexical embeddings; see
+            # https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/pytorch_pretrained_bert/modeling.py#L186
+            h_lex = self.model.embeddings.word_embeddings(ids)
+            h_lex = self.model.embeddings.LayerNorm(h_lex)
+            # following our use of the OpenAI model, don't use dropout for
+            # probing.
+            #  h_lex = self.model.embeddings.dropout(embeddings)
```
Comment (W4ngatang) [inline src/bert/utils.py]: Maybe make it an option whether or not the model is being used in a probing context? I think we'll (read: I will want to) use this in a non-probing context.
Also, I forget is the convention is to apply dropout at the end of a module or at the beginning of the next use (I remember this caused an issue during the workshop).

File: src/bert/utils.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+        if self.embeddings_mode != "only":
+            # encoded_layers is a list of layer activations, each of which is
+            # <float32> [batch_size, seq_len, output_dim]
+            encoded_layers, _ = self.model(ids, token_type_ids=torch.zeros_like(ids),
```
Comment (W4ngatang) [inline src/bert/utils.py]: does this do double computation of embedding layers if you're in `cat` mode? If so, maybe reuse `h_lex` if it's there? (though I guess lookups are fast).

File: src/models.py [hunk 7]
```diff
@@ -69,6 +69,21 @@ def build_model(args, vocab, pretrained_embs, tasks):
...
+        # should prevent these from being enabled anyway.
+        from .bert.utils import BertEmbedderModule
+        log.info(f"Using BERT model ({args.bert_model_name}); skipping other embedders.")
+        cove_layer = None
+        # Set PYTORCH_PRETRAINED_BERT_CACHE environment variable to an existing
+        # cache; see
+        # https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/pytorch_pretrained_bert/file_utils.py
+        bert_cache_dir = (os.path.join(args.exp_dir, "bert_cache") if
+                          args.bert_fine_tune else None)
+        if bert_cache_dir:
```
Comment (W4ngatang) [inline src/models.py]: dumb question: why only cache if fine tuning? Is the cache dir for storing fine tuned weights?

File: src/preprocess.py [hunk 8]
```diff
@@ -259,6 +262,13 @@ def build_tasks(args):
                              ("OpenAI transformer is not supported alongside"
                               " other indexers due to tokenization!")
         indexers["openai_bpe_pretokenized"] = SingleIdTokenIndexer("openai_bpe")
+    if args.bert_model_name:
+        assert not indexers, ("BERT is not supported alongside"
+                              " other indexers due to tokenization!")
+        assert args.tokenizer == args.bert_model_name, \
+                             ("BERT is not supported alongside"
+                              " other indexers due to tokenization!")
```
Comment (W4ngatang) [inline src/preprocess.py]: should be "tokenizers"?

File: src/bert/utils.py (hunk 3 above)
Comment (iftenney) [reply src/bert/utils.py]: I don't think so, < 512 should also work since BERT has an internal attention mask (https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/pytorch_pretrained_bert/modeling.py#L620)

File: src/bert/utils.py (hunk 4 above)
Comment (iftenney) [reply src/bert/utils.py]: Good catch. We pass the tokens through the BERT WordpieceTokenizer, which checks for OOV and replaces with it's own unk token https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/pytorch_pretrained_bert/tokenization.py#L329 (which has some id > 1 in the main vocab), so this should never happen.

Added an assert to complain in case something goes wrong.

File: src/bert/utils.py (hunk 5 above)
Comment (iftenney) [reply src/bert/utils.py]: FWIW, this is basically a shortcut to treating the BERT WPM like GloVe vectors. Erring on the side of no dropout here, since there are a bunch of places we can add it later if these are the only embeds?

I'd rather not add an additional "is_probing" param since that introduces a lot of cruft.

File: src/bert/utils.py (hunk 6 above)
Comment (iftenney) [reply src/bert/utils.py]: Yes it's a little wasteful but I'd have to break open the BertModel implementation to re-use h_lex. As implemented here we don't have to modify that code at all.

File: src/models.py (hunk 7 above)
Comment (iftenney) [reply src/models.py]: Good point; I think this is unnecessary. Safest thing is to download once then mark as read-only, I guess?

There was a line in the documentation advising to use separate directories for each worker in distributed training, but this might just be to avoid hammering a single file server.

File: src/preprocess.py (hunk 8 above)
Comment (iftenney) [reply src/preprocess.py]: Fixed.

File: src/models.py (hunk 7 above)
Comment (iftenney) [reply src/models.py]: Removed this special case; should always use cache if the environment var is set.

File: src/bert/utils.py (hunk 3 above)
Comment (W4ngatang) [reply src/bert/utils.py]: :+1:

File: src/bert/utils.py (hunk 4 above)
Comment (W4ngatang) [reply src/bert/utils.py]: sounds good, nit: should add a more descriptive error message here

File: src/bert/utils.py (hunk 5 above)
Comment (W4ngatang) [reply src/bert/utils.py]: :+1:

**Comment by sleepinyourhat** [comment]
Ready to merge?

**Comment by sleepinyourhat** [comment]
Nudge again. Other stuff is waiting on this.
